clang/lib/Format/TokenAnnotator.cpp

```diff
@@ -243,14 +244,16 @@ class AnnotatingParser {
       // operator that was misinterpreted because we are parsing template
       // parameters.
       // FIXME: This is getting out of hand, write a decent parser.
-      if (InExpr && !Line.startsWith(tok::kw_template) &&
+      if (InExpr && !SeenFatArrow && !Line.startsWith(tok::kw_template) &&
           Prev.is(TT_BinaryOperator)) {
         const auto Precedence = Prev.getPrecedence();
         if (Precedence > prec::Conditional && Precedence < prec::Relational)
           return false;
       }
       if (Prev.isOneOf(tok::question, tok::colon) && !Style.isProto())
         SeenTernaryOperator = true;
+      else if (Prev.is(TT_FatArrow))
```

Comment:
i am not sure if the regression is specific to fat-arrow here (sorry for not putting much effort into the reproducer). the original change seem to changed behavior for handling of any bitwise operators. so I think something like:
```cpp
auto foo = new Bar<kFoo | kBar>();
```

would also be formatted differently

Reply:
How realistic is your C++ example? In general, I don't think we can really tell if the angle brackets are template opener/closer or comparison operators without checking more (and very specific) contexts, e.g. the `new` keyword before `Bar` and the empty parentheses after `>`. In absence of that, it seems more likely that the angles are comparison operators. See e.g. this [test case](https://github.com/llvm/llvm-project/blob/95a0b4f729310d95d89f01f4d92ab7d2bf09941c/clang/unittests/Format/TokenAnnotatorTest.cpp#L645).

Reply:
i don't think new or empty parantheses are a must either, the logic around here is just checking that `<>` are inside an expression, hence that kind of code pattern is likely quite common as people tend to do arithmetic via metaprogramming using these patterns. e.g:
```cpp
constexpr auto foo() { return FixedInt<N | M>(foo); }
```

this will be formatted as the following now:
```cpp
constexpr auto foo() { return FixedInt < N | M > (foo); }
```

Reply:
Without extra contexts, how can you tell whether the less/greater below are comparison operators or angle brackets?
```
return a < b | c > (d ^ e);
```

Reply:
sorry I wasn't saying that `FixedInt<N | M>(foo);` is clearly a template argument, it's definitely ambiguous in absence of name-resolution and there isn't much clang-format can do here.

this kind of backward incompatible changes are making it harder to update clang-format to newer versions in large codebases. as clang-format right now is trading off one set of false-formatting with other. hence my understanding for the project has been to `stick with existing formatting, if we're heuristically detecting some patterns. so that behavior on existing formatted-code doesn't regress.` maybe that understanding has changed over time, or wasn't there at all. But this bug and requested changes in your PR was solely made with that assumption (as many others I've been filing over the past years).

hence I am not expecting a "fix" for the issue here, but I guess a narrowed down version of [other patch](https://github.com/llvm/llvm-project/commit/73c961a3345c697f40e2148318f34f5f347701c1), where it actually preserves behavior with old clang-format in uncertain cases.

Reply:
> hence I am not expecting a "fix" for the issue here, but I guess a narrowed down version of [other patch](https://github.com/llvm/llvm-project/commit/73c961a3345c697f40e2148318f34f5f347701c1), where it actually preserves behavior with old clang-format in uncertain cases.

This patch IMO *is* a narrowed down version of the other patch #100980, which fixed a real regression #100300 caused by yet another patch #96801, which in turn was an attempt at fixing a real bug #81385.

In your [reply](https://github.com/llvm/llvm-project/pull/108671#discussion_r1762725421) above, it doesn't seem that your C++ example `return FixedInt<N | M>(foo);` is from an existing codebase. The fact that it happened to get formatted that way before doesn't necessarily mean that we should always format another similar pattern e.g. `return Foo < A || B > (C ^ D);` the same way, especially if the latter is much more likely.

Reply:
I've added an option in #109916 to help disambiguate the `<` in the C++ examples above.

Reply:
> This patch IMO is a narrowed down version of the other patch https://github.com/llvm/llvm-project/pull/100980, which fixed a real regression https://github.com/llvm/llvm-project/issues/100300 caused by yet another patch https://github.com/llvm/llvm-project/pull/96801, which in turn was an attempt at fixing a real bug https://github.com/llvm/llvm-project/issues/81385.

I totally get that. The point I am trying to make is, these sequences of patches, fixed a real regression, at the cost of introducing another one.

> In your https://github.com/llvm/llvm-project/pull/108671#discussion_r1762725421 above, it doesn't seem that your C++ example return FixedInt<N | M>(foo); is from an existing codebase.

Sorry if I gave that impression, but I was delibaretly trying to make it more concrete as you were asking for more specifics, here's some real world examples: https://github.com/search?q=lang%3Ac%2B%2B+%2FFixedInt%3C.*%5C%7C.*%3E%2F&type=code

> The fact that it happened to get formatted that way before doesn't necessarily mean that we should always format another similar pattern e.g. return Foo < A || B > (C ^ D); the same way, especially if the latter is much more likely.

No I totally get that. That's how clang-format evolves over time. But as I explained above, my understanding is this is done in a way that it doesn't regress behavior for already formatted code. As otherwise it becomes really troublesome to adopt new versions of clang-format.
Hence the change in https://github.com/llvm/llvm-project/pull/100980, didn't simply improve formatting for some mishandled pattern, it also regressed handling of some existing pattern.

I am totally OK if your verdict here is this is a cost we're willing to pay, in the end you're the maintainer. I am just trying to make sure this doesn't go unnoticed and someone is making a call deliberately. Even if it isn't aligned with the outcome that I might like.

> I've added an option in https://github.com/llvm/llvm-project/pull/109916 to help disambiguate the < in the C++ examples above.

It might be a nice addition, but it doesn't really change the fact that clang-format-18 is WAI for such code patterns, and starting with clang-format-19 this has regressed and users need to change their configuration. Moreover it also means that patch now needs to be backported (or clang-format-19 is just broken for certain codebases).

Reply:
> > In your [#108671 (comment)](https://github.com/llvm/llvm-project/pull/108671#discussion_r1762725421) above, it doesn't seem that your C++ example return FixedInt<N | M>(foo); is from an existing codebase.
> 
> Sorry if I gave that impression, but I was delibaretly trying to make it more concrete as you were asking for more specifics, here's some real world examples: https://github.com/search?q=lang%3Ac%2B%2B+%2FFixedInt%3C.*%5C%7C.*%3E%2F&type=code

The GitHub examples you listed here have zero arguments and thus are unambiguous to clang-format. (See #110408.) As I mentioned or implied above, a pair of empty parentheses after the `>` would help. We can add another patch (if needed) for multiple arguments such as `return FixedInt<N | M>(foo, bar);` , which is also unambiguous.

However, the initial version of your C++ example `return FixedInt<N | M>(foo);` has exactly one argument and can be also interpreted as a non-template expression similar to `Foo < A || B > (C ^ D);`.

> > The fact that it happened to get formatted that way before doesn't necessarily mean that we should always format another similar pattern e.g. return Foo < A || B > (C ^ D); the same way, especially if the latter is much more likely.
> 
> No I totally get that. That's how clang-format evolves over time. But as I explained above, my understanding is this is done in a way that it doesn't regress behavior for already formatted code. As otherwise it becomes really troublesome to adopt new versions of clang-format. Hence the change in #100980, didn't simply improve formatting for some mishandled pattern, it also regressed handling of some existing pattern.

It goes without saying that we should strive to keep clang-format regression free. Nevertheless, not all behavior changes are regressions. Also, some past regressions can't simply be reverted as it would cause new regressions. The change in #100980 is an example of the latter.

> I am totally OK if your verdict here is this is a cost we're willing to pay, in the end you're the maintainer. I am just trying to make sure this doesn't go unnoticed and someone is making a call deliberately. Even if it isn't aligned with the outcome that I might like.

With #109916 we can cover both `return Foo < A || B > (C ^ D);` and `return Foo<A | B>(C ^ D);`. It may require changing the existing configuration for the latter case, but the alternative (i.e. not supporting the former case or having a `NonTemplateNames` option instead) would be worse IMO.

This patch is to fix the TypeScript issue as initially reported in #108536. As #109925 also fixes that issue (probably incidentally), I'll close this one if you think special casing the fat arrow for JavaScript/TypeScript is unnecessary.

Reply:
> The GitHub examples you listed here have zero arguments and thus are unambiguous to clang-format. (See https://github.com/llvm/llvm-project/pull/110408.) As I mentioned or implied above, a pair of empty parentheses after the > would help. We can add another patch (if needed) for multiple arguments such as return FixedInt<N | M>(foo, bar); , which is also unambiguous.

Sorry I wasn't paying attention to such details, as I was rather trying to explain the current logic and the previous one has different characteristics around `<>`. Trying to fix it for certain special cases won't change that fact and we'll let regressions around cases that we haven't discovered to slip.

Hence, I totally agree with your approach on only changing behavior in special cases, rather than changing overall behavior. The point that's puzzling me is, why don't we treat 834ac2e205dd8e492d6084a7952e68e19a1f54db and 73c961a3345c697f40e2148318f34f5f347701c1 similarly. If they were to just change handling of ternary expressions, but they changed behavior for expressions that don't have `?` in them, it sounds like an unintended change.

> However, the initial version of your C++ example return FixedInt<N | M>(foo); has exactly one argument and can be also interpreted as a non-template expression similar to Foo < A || B > (C ^ D);.

I completely agree with that, and that's the regression I am trying to express as well. Previously clang-format was erring towards recognizing these as template arguments, now it's erring towards recognizing them as comparisons. I am trying to figure out if we share the same understanding on this topic. If we both agree on this one, but you folks believe that's a regression we're willing to eat, I am totally fine with that. In that case I think https://github.com/llvm/llvm-project/pull/109916 LGTM!

Reply:
> If they were to just change handling of ternary expressions, but they changed behavior for expressions that don't have ? in them, it sounds like an unintended change.

All these template edge-cases weren't being represented in testing. My change broke no tests, but it caused a lot of fallout in the issues. I regret making that change, but by now it's too late. I'm sorry

Reply:
> The point that's puzzling me is, why don't we treat https://github.com/llvm/llvm-project/commit/834ac2e205dd8e492d6084a7952e68e19a1f54db and https://github.com/llvm/llvm-project/commit/73c961a3345c697f40e2148318f34f5f347701c1 similarly.

The latter was an attempt at fixing both the regression caused by the former and the [issue](https://github.com/llvm/llvm-project/issues/100300#issuecomment-2249303341) with bitwise operators _outside_ templates. By the time the issue with bitwise operators _inside_ templates was [brought up](https://github.com/llvm/llvm-project/pull/108671#discussion_r1760689981), 392b77d58a91049a155f3390ec16941a848aa766 and another commit d9806ffe4e4d26de9c01f6b8ac0deae169b1d88d on top of it had already been released in 19.1.0.

Reply:
It seems a general rule of thumb that the more complicated the code the harder the job clang-format seems to have.

Whilst I admire the programming prowess of such code, I generally struggle to read it no matter how its formatted. My view has often been when clang-format does fail, it might say more about the code than about clang-format. For the cases we butcher (and we will from time to time) there is still //clang-format off until we can resolve it.

that said I believe we can only walk towards clang-format doing the right thing one patch at a time, and I don't believe we should let "Perfection be the enemy of good". 

I personally like to follow the "Beyonce" rule, "if you liked it you should have put a test on it". IMHO as long as our tests pass, and the test cases keep growing, and we don't allow tests to change without good reason, and all new code has tests, then I believe we are good. 

Some person, with some specific failure will always happen, given how much code is likely clang-formatted daily. But honestly I don't think of it as a true regression unless the test is failing, because often regressions are caused by
us now interpreting better, annotating better, or just fixing something and often what previously worked by chance before doesn't now. That's not a real regression its just bad luck and a lack of a test to keep us honest.

While I don't really like code becoming a glorious game of whack-a-mole I think our intentions are honorable and we should make forward process, even if its one patch at a time.

When this does happen it means that previous use case wasn't covered by a test. It needs to be, and we can fix it accordingly adding the missing test which ultimately should give us more robustness around the unintended "bad-luck"

However the clang-format users community must remember, we are a core team of 4 with really @owenca doing what I consider to be the majority share, (sorry I'm so busy with my day job. I only get time for drive by reviewing, but I'm trying to continue to give moral support), as far as I am aware none of us are paid to work on clang-format, and despite all the big IDE players using clang-format integrated into their commercial offerings I see little in the way of contribution from them. In fact at last years CppCon I saw one even openly criticize clang-format with vague obtuse corner cases (A cheap shot in my view).

Its my belief that "our modus operandi" is as correct as it can be given our time and resources. We never intentionally break things, but its extremely hard to find a large enough code base that we can test against especially given that not even all of LLVM or even Clang is clang-format clean, enough to uncover issues.  This is especially difficult to find code using new C++ language concepts, modules, complex template scenarios and lambdas.  

Now if anyone wants to persuade all of  LLVM that the entire repo should be clang-formatted top to bottom, then maybe we'd have more code examples to ensure we introduced less issues.

Reply:
If you are interested in the "Cheap Shot" ... https://www.youtube.com/watch?v=NnQraMtpvws This is from a JetBrains employee which frankly I'd have expected better from given the feedback I gave them on their blog  https://blog.jetbrains.com/clion/2023/12/a-clangformat-story-and-the-third-clion-nova-update/